### PR TITLE
Close db with open iterator should not crash

### PR DESCRIPTION
**Pull request by peakji**
Discussions: https://github.com/Level/leveldown/issues/194

### CODE REVIEW

**Comment by vweevers** [comment]
This means the JS callback won't be called. Is that desirable? Let's make an explicit choice:

1. limbo
2. call with an error (something like "iterator has ended")
3. call with `finished = true` (ending the JS iterator)

WDYT? @ralphtheninja @juliangruber

**Comment by peakji** [comment]
@vweevers I think 2 is the best.

> 1. limbo

I implemented the current solution (limbo) because an iterator can only be marked as `ended` with **Iterator::End**, which can be triggered:

* By the user with `iterator.end()`. We don't expect users to call `iterator.next()` after `iterator.end()`.
* By the auto-close code in `database.close()`. I think its fine to ignore the callback of `iterator.next()`, because the code should continue from the callback of `database.close()`.

This is an aggressive solution and might result in unexpected behavior.

> 2. call with an error (something like "iterator has ended")

The only concern is, should we give the error after `database.close()`? Its actually not an error in this case, and may trigger the users error handling code for normal iterator errors.

> 3. call with `finished = true` (ending the JS iterator)

Some users use `finished` to check iterator status, e.g. calculate progress by counting `finished` in an array of iterators. If we set `finished = true` here, it may give the illusion of all the tasks were **completed**. 

Or we could renamed it into something like `terminated`.

**Comment by juliangruber** [comment]
hm, I still don't have need for `db.close()`, in all my use cases closing the process explicitly was enough.

File: src/iterator.cc [hunk 1]
```diff
@@ -323,15 +323,16 @@ NAN_METHOD(Iterator::Seek) {
...
-    return;
-
   if (!info[0]->IsFunction()) {
     return Nan::ThrowError("next() requires a callback argument");
   }
 
   v8::Local<v8::Function> callback = info[0].As<v8::Function>();
 
+  if (iterator->ended) {
+    LD_RETURN_CALLBACK_OR_ERROR(callback, "iterator has ended");
```
Comment (ralphtheninja) [inline src/iterator.cc]: 👍

**Comment by ralphtheninja** [comment]
I prefer 2.

Great work @peakji

**Comment by peakji** [comment]
@juliangruber We're trying to solve a crash caused by `db.close()`, details: https://github.com/Level/leveldown/issues/194#issuecomment-306686747.
It seems [pouchdb](https://github.com/pouchdb/pouchdb) is using this feature.

**Comment by ralphtheninja** [comment]
This will solve many problems and we should be able to close several issues in the backlog.

File: test/iterator-test.js [hunk 2]
```diff
@@ -152,6 +152,24 @@ make('iterator seek before next has completed', function (db, t, done) {
...
 })
 
+make('close db with open iterator', function (db, t, done) {
+  var ite = db.iterator()
+  var cnt = 0
+  ite.next(function loop(err, key, value) {
+    if(cnt++ === 0)
+      t.error(err, 'no error from next()')
+    else
+      t.ok(err, 'had error from next() after end()')
```
Comment (ralphtheninja) [inline test/iterator-test.js]: Maybe check the error message here instead, e.g. `t.equal(err.message, 'iterator has ended')`

Comment (ralphtheninja) [inline test/iterator-test.js]: cc @peakji

Comment (peakji) [inline test/iterator-test.js]: 👌

**Comment by ralphtheninja** [comment]
LGTM!

**Comment by peakji** [comment]
close https://github.com/Level/leveldown/issues/194

**Comment by ralphtheninja** [comment]
1.7.2

**Pull request merged**